c_cpp_guide.md

```diff
@@ -0,0 +1,117 @@
+- The module's exports must be a `main` function which takes no arguments and returns nothing, and the `memory` of the module.
+- The module may not use floats or other [sources of non-determinism](https://github.com/WebAssembly/design/blob/master/Nondeterminism.md).
+
+## Caveats
+
+When writing Ewasm contracts in C/C++, one should bear in mind the following caveats:
+
+1. WebAssembly is still primitive and [lacks features](https://github.com/WebAssembly/design/blob/master/FutureFeatures.md). For example, WebAssembly lacks support for exceptions and we have no way to do system calls in Ewasm. Compilers and libraries are still primitive. For example, we have a patched version of libc to allow `malloc`, but the patches are not yet enough for `std::vector` because other memory managment calls are unavailable. But perhaps any memory management beyond memory allocation may be unwanted for Ewasm contracts since it costs gas. This situation will improve as WebAssembly, compilers, and libraries mature.
+
+1. In the current Ewasm design, all communication between the contract and the client is done through the module's memory. For example, the message data ("call data") sent to the contract is accessed by calling `callDataCopy()`, which puts this data to WebAssembly memory at a location given by a pointer. This pointer must be to either to a statically allocated array, or to dynamically allocated memory using `malloc`. For example, before calling `callDataCopy()`, one may use `getCallDataSize()` to see how many bytes of memory to `malloc`.
+
+1. In the current Ewasm design, the Ethereum client writes data into WebAssembly as big-endian, but WebAssembly memory is little-endian, so has reversed bytes when the data is brought to/from the WebAssembly operand stack. For example, when the call data is brought into memory using `callDataCopy`, and those bytes are loaded to the WebAssembly stack using `i64.load`, all of the bytes are reversed. So extra C/C++ code may be needed to load bytes from the correct location and to reverse the loaded bytes.
+
+1. The output of compilers is a `.wasm` binary which may have imports and exports which do not meet Ewasm requirements. We have tools to fix the imports and exports.
+
+1. There are no tutorials for debugging/testing a contract. Hera supports extra Ewasm helper functions to print things, which have helped in writing test cases. A tutorial is needed to allow early adopters to debug/test their contracts without having to do it on the testnet.
+
+## Basic Step-by-Step Guide
+
+First build the latest version of LLVM.
+
+Aside: If you wish to use C/C++ standard libraries, then build the version of LLVM in the Advanced section below. That version can also be used here.
+
+```sh 
+# checkout LLVM, clang, and lld
+svn co http://llvm.org/svn/llvm-project/llvm/trunk llvm
```

Comment:
Can we use git instead of svn? The instructions from [Jake's doc](https://github.com/ewasm/testnet/blob/7742ce13586e386cd12f7d669a0dde294a8b6966/clang.md#from-the-repo) seem reasonable no?

Reply:
I mentioned that the official guide http://llvm.org/docs/GettingStarted.html uses svn. But changed our guide to use git.

Reply:
I am still compiling the git version to test it. Will revert to svn if there is a problem.

Reply:
The git version successfully compiles wrc20. Compiling the git version of LLVM had a few errors along the way, but restarted each time and finally it finished.

Reply:
Didn't realize svn was in the official guide. Glad to hear it works with git too!